src/Compilers/Core/Portable/DiagnosticAnalyzer/DefaultAnalyzerAssemblyLoader.Core.cs

```diff
@@ -2,45 +2,102 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
-#nullable disable
+#nullable enable
 
 #if NETCOREAPP
 
-using System.Diagnostics;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
 using System.Reflection;
 using System.Runtime.Loader;
 
 namespace Microsoft.CodeAnalysis
 {
     internal class DefaultAnalyzerAssemblyLoader : AnalyzerAssemblyLoader
     {
-        private AssemblyLoadContext _loadContext;
+        private readonly object _guard = new object();
+        private readonly Dictionary<string, AssemblyLoadContext> _loadContextByDirectory = new Dictionary<string, AssemblyLoadContext>();
 
         protected override Assembly LoadFromPathImpl(string fullPath)
         {
-            //.NET Native doesn't support AssemblyLoadContext.GetLoadContext. 
-            // Initializing the _loadContext in the .ctor would cause
-            // .NET Native builds to fail because the .ctor is called. 
-            // However, LoadFromPathImpl is never called in .NET Native, so 
-            // we do a lazy initialization here to make .NET Native builds happy.
-            if (_loadContext == null)
+            if (!ShouldLoad(fullPath))
             {
-                AssemblyLoadContext loadContext = AssemblyLoadContext.GetLoadContext(typeof(DefaultAnalyzerAssemblyLoader).GetTypeInfo().Assembly);
+                throw new ArgumentException(message: null, paramName: nameof(fullPath));
+            }
+
+            AssemblyLoadContext? loadContext;
 
-                if (System.Threading.Interlocked.CompareExchange(ref _loadContext, loadContext, null) == null)
+            var fullDirectoryPath = Path.GetDirectoryName(fullPath) ?? throw new ArgumentException(message: null, paramName: nameof(fullPath));
+            lock (_guard)
+            {
+                if (!_loadContextByDirectory.TryGetValue(fullDirectoryPath, out loadContext))
                 {
-                    _loadContext.Resolving += (context, name) =>
-                    {
-                        Debug.Assert(ReferenceEquals(context, _loadContext));
-                        return Load(name.FullName);
-                    };
+                    loadContext = new DirectoryLoadContext(fullDirectoryPath, this);
```

Comment:
I think this has an issue where the "original" assembly path is getting mixed up with the "shadow copy" dependency path. i.e. we are registering the "original" path with AddDependencyLocation, then trying to load it via its shadow path and rejecting it. Will have to figure out how to rationalize all this.